src/genome.py

```diff
@@ -88,154 +97,260 @@ def __str__(self) -> str:
+
+def insert_next(link: Link[T], val: T) -> None:
+    """Add a new link containing avl after link."""
+    new_link = Link(val, link, link.next)
+    new_link.prev.next = new_link
+    new_link.next.prev = new_link
+
+Ft = namedtuple("Ft", ["ft", "length"])
 
 
 class LinkedListGenome(Genome):
     """
     Representation of a genome.
-
     Implements the Genome interface using linked lists.
     """
+    head: Link[Ft]  
+    active = {}
+    counter = 0
 
     def __init__(self, n: int):
         """Create a new genome with length n."""
-        ...  # FIXME
+        self.head = Link(None, None, None)
+        self.head.next = self.head
+        self.head.prev = self.head
+        insert_next(self.head.prev, Ft(0, n))
 
     def insert_te(self, pos: int, length: int) -> int:
         """
         Insert a new transposable element.
-
         Insert a new transposable element at position pos and len
         nucleotide forward.
-
         If the TE collides with an existing TE, i.e. genome[pos]
         already contains TEs, then that TE should be disabled and
         removed from the set of active TEs.
-
         Returns a new ID for the transposable element.
         """
-        ...  # FIXME
-        return -1
+        for ft, end in self.iter_pos():
+            if end <= pos:
+                continue
+            else:
+                half_1 = ft.val.length -end + pos
+                half_2 = end - pos
+                self.insert_te_at_pos(length, ft, half_1, half_2)
+                return self.counter
+        return None
+
+    def insert_te_at_pos(self, length, ft, half_1, half_2):
+        if  ft.val.ft == 1:
+            self.disable_feature(ft)
+        insert_next(ft, Ft(1, length))
+        ft.val = Ft(ft.val.ft, half_1)
+        insert_next(ft.next, Ft(ft.val.ft, half_2))
+        self.counter += 1
+        self.active[self.counter] = ft.next
+
+    def disable_feature(self, ft):
+        ft.val = Ft(2, ft.val.length)
+        for identifier, disable in self.active.items():
+            if disable is ft:
+                self.active.pop(identifier)
+                break
+
 
     def copy_te(self, te: int, offset: int) -> int | None:
         """
         Copy a transposable element.
-
         Copy the transposable element te to an offset from its current
         location.
-
         The offset can be positive or negative; if positive the te is copied
         upwards and if negative it is copied downwards. If the offset moves
         the copy left of index 0 or right of the largest index, it should
         wrap around, since the genome is circular.
-
         If te is not active, return None (and do not copy it).
         """
-        ...  # FIXME
+
+        ft = self.active[te]
+        temp = self.active[te]
+        
+        if temp.val.ft != 1:
+            return None
+
+        if offset > 0:
+            offset = offset - temp.val.length
+            while offset > 0:
+                ft = getattr(ft, "next")
```

Comment:
Why on earth do you use `getattr` here? Don't get me wrong, the function has its usage, but using `getattr(ft,"next")` when `ft.next` does the exact same thing seems strange.